### PR TITLE
File descriptor cache for fileio daemon [ESD-119]

### PR DESCRIPTION
**Pull request by silverjam**
Instead of re-opening and closing the file descriptor for each ~240b write/read request, keep the file descriptor open and cache it.  If another request comes in within 3s for the same file descriptor (in the same mode) then we'll re-use the already open descriptor.

Initial results show that this results in about a 20-35% speed-up.  Without the cache:

screen shot 2018-12-20 at 4 57 32 pm

With the cache:

screen shot 2018-12-20 at 4 47 54 pm

### CODE REVIEW

**Comment by silverjam** [comment]
Naively sampling the `sbp_fileio` metrics with the following snippet:
```
watch 'find /var/log/metrics/sbp_fileio -type f -print -exec cat {} \;'
```

**Comment by silverjam** [comment]
Two remaining issues:

+ This doesn't currently write data correctly, probably because we're calling lseek on the file descriptor underlying the `FILE*`-- and each time we're not actually writing data but just seeking the FD-- when the internal buffer gets flushing we've just been seeking farther and farther out.

This seems corroborated by this:

screen shot 2018-12-20 at 5 45 39 pm

Which shows that the front of the file is filled with zeros

+ Need to figure out how to flush the file descriptors when something else wants to read a while, e.g. the upgrade process needs to tell the daemon to flush its buffers because it needs to verify the whole content of the file.

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v2.1.0-develop-2018121420-16-g44df8a71

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=820879)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail"
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2018121420-16-g44df8a71&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2018121420-16-g44df8a71&build_type=buildroot_pr&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v2.1.0-develop-2018121420-16-g44df8a71
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2018121420-16-g44df8a71)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2018121420-16-g44df8a71/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2018121420-16-g44df8a71

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v2.1.0-develop-2018121420-17-g860d5ce4

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=820921)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail"
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2018121420-17-g860d5ce4&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2018121420-17-g860d5ce4&build_type=buildroot_pr&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v2.1.0-develop-2018121420-17-g860d5ce4
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2018121420-17-g860d5ce4)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2018121420-17-g860d5ce4/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2018121420-17-g860d5ce4

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c [hunk 1]
```diff
@@ -73,12 +320,120 @@ void sbp_fileio_setup(path_validator_t *pv_ctx,
...
+  struct stat wait_stat;
+  if (stat(sbp_fileio_wait_flush_file, &wait_stat) < 0) {
+    piksi_log(LOG_ERR,
+              "calling stat on flag file '%s' failed: %s",
+              sbp_fileio_wait_flush_file,
+              strerror(errno));
+    return false;
+  }
+
+  pid_t pid = (pid_t)atoi(pid_buffer);
```
[the remark this answers is not in the document]
Comment (silverjam) [reply package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c]: 👍

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/flush_main.c [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  puts("Flush any cached data in a sbp_fileio_daemon to disk");
+  puts("");
+  puts("-h, --help            Print this message");
+  puts("-d, --debug           Output debug logging");
+  puts("-n, --name            The name of the daemon to flush");
+}
+
+static int parse_options(int argc, char *argv[])
+{
+  // clang-format off
```
[the remark this answers is not in the document]
Comment (silverjam) [reply package/sbp_fileio_daemon/sbp_fileio_daemon/src/flush_main.c]: 👍

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c [hunk 3]
```diff
@@ -22,23 +22,73 @@
...
 #include <libpiksi/util.h>
-#include <libsbp/file_io.h>
 
 #include "sbp_fileio.h"
 #include "path_validator.h"
 
 #include "fio_debug.h"
 
+#define WAIT_FLUSH_RETRY_MS 50
+#define MS_TO_NS(MS) ((MS)*1e6)
```
[the remark this answers is not in the document]
Comment (silverjam) [reply package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c]: 👍

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/flush_main.c [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+#include <libpiksi/logging.h>
+#include <libpiksi/util.h>
+#include <libpiksi/sbp_pubsub.h>
+
+#include "sbp_fileio.h"
+#include "fio_debug.h"
+
+#define PROGRAM_NAME "sbp_fileio_flush"
+
+bool fio_debug = false;
```
[the remark this answers is not in the document]
Comment (silverjam) [reply package/sbp_fileio_daemon/sbp_fileio_daemon/src/flush_main.c]: these are extern'd, I moved these to a common location and added a comment to indicate that they're extern

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c [hunk 5]
```diff
@@ -53,17 +103,214 @@ static void read_dir_cb(u16 sender_id, u8 len, u8 msg[], void *context);
...
+static void purge_fd_cache(pk_loop_t *loop, void *handle, int status, void *context);
+static void flush_cached_fd(const char *path, const char *mode);
+static fd_cache_result_t open_from_cache(const char *path,
+                                         const char *mode,
+                                         int oflag,
+                                         mode_t perm);
+
+static bool test_control_dir(const char *name);
+
+static fd_cache_t fd_cache[FD_CACHE_COUNT] = {[0 ...(FD_CACHE_COUNT - 1)] = {NULL, "", "", 0, 0}};
```
[the remark this answers is not in the document]
Comment (silverjam) [reply package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c]: 👍

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c [hunk 6]
```diff
@@ -73,12 +320,120 @@ void sbp_fileio_setup(path_validator_t *pv_ctx,
...
+
+  piksi_log(LOG_INFO, "sbp_fileio_daemon flush for named daemon '%s' done", name);
+  return success;
 }
 
 static int allow_mtd_read(const char *path)
 {
-  // TODO/DAEMON-USERS: Replace this with an SBP message (give firmware a handle
-  //    to read here instead of allowing the read).
+  // TODO/DAEMON-USERS: Replace this with an SBP message, e.g. give the firmware a handle
```
[the remark this answers is not in the document]
Comment (silverjam) [reply package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c]: 👍

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c [hunk 7]
```diff
@@ -73,12 +320,120 @@ void sbp_fileio_setup(path_validator_t *pv_ctx,
...
+  int rc = unlink(sbp_fileio_wait_flush_file);
+  if (rc < 0) {
+    piksi_log(LOG_ERR,
+              "unlink of flag file '%s' failed: %s",
+              sbp_fileio_wait_flush_file,
+              strerror(errno));
+  }
+}
+
+bool sbp_fileio_request_flush(const char *name)
```
[the remark this answers is not in the document]
Comment (silverjam) [reply package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c]: split into several helper functions

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.h [hunk 8]
```diff
@@ -18,10 +18,45 @@
...
+
+/**
+ * Setup the process for operation of the fileio daemon.
+ *
+ * @param name    the name of the FileIO daemon (e.g. internal / external) used to name
+ *                metrics and control files.
+ * @param loop    the loop that will handle fileio requests
+ * @param pv_ctx  the @see path_validator_t object that will check if a path is allowed for fileio
+ * @param allow_factory_mtd
+ *    if pathes from /factory are allowed for fileio requests
```
[the remark this answers is not in the document]
Comment (silverjam) [reply package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.h]: Fixed

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v2.1.0-develop-2019010420-19-gabf450ae

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=850951)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail"
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019010420-19-gabf450ae&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019010420-19-gabf450ae&build_type=buildroot_pr&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v2.1.0-develop-2019010420-19-gabf450ae
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2019010420-19-gabf450ae)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2019010420-19-gabf450ae/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2019010420-19-gabf450ae

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v2.1.0-develop-2019010420-20-gc2641e84

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=851075)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail"
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019010420-20-gc2641e84&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019010420-20-gc2641e84&build_type=buildroot_pr&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v2.1.0-develop-2019010420-20-gc2641e84
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2019010420-20-gc2641e84)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2019010420-20-gc2641e84/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2019010420-20-gc2641e84

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c [hunk 9]
```diff
@@ -130,20 +491,41 @@ static void read_cb(u16 sender_id, u8 len, u8 msg_[], void *context)
...
-    close(f);
+
+    flush_cached_fd(msg->filename, "w");
+    cached_fd = open_from_cache(msg->filename, "r", O_RDONLY, 0);
+
+    if (cached_fd.fp != NULL) {
+
+      if (read_offset(cached_fd) != (off_t)msg->offset) {
+        seek_result = fseek(cached_fd.fp, msg->offset, SEEK_SET);
+        if (seek_result == 0) {
```
[the remark this answers is not in the document]
Comment (silverjam) [reply package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c]: I removed 1 level of nesting with a `goto`... I hope you're happy 😛

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v2.1.0-develop-2019010420-21-ga45c8520

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=851390)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail"
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019010420-21-ga45c8520&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019010420-21-ga45c8520&build_type=buildroot_pr&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v2.1.0-develop-2019010420-21-ga45c8520
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2019010420-21-ga45c8520)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2019010420-21-ga45c8520/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2019010420-21-ga45c8520

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (pmiettinen) [reply package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c]: I can survive 3 every now and then.

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.h [hunk 10]
```diff
@@ -18,10 +18,68 @@
...
+extern bool no_cache;
+
+/**
+ * Setup the process for operation of the fileio daemon.
+ *
+ * @param name                The name of the FileIO daemon (e.g. internal / external) used to name
+ *                            metrics and control files.
+ * @param loop                The loop that will handle fileio requests
+ * @param pv_ctx              The @c path_validator_t object that will check if a path is allowed
+ * for fileio
```
Comment (pmiettinen) [inline package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.h]: formatting

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c [hunk 11]
```diff
@@ -256,12 +710,13 @@ static void remove_cb(u16 sender_id, u8 len, u8 msg[], void *context)
 static void write_cb(u16 sender_id, u8 len, u8 msg_[], void *context)
 {
   (void)sender_id;
+  u8 headerlen = 0;
```
Comment (pmiettinen) [inline package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c]: This definition could be closer to the place of use like other variables.

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c [hunk 12]
```diff
@@ -53,32 +103,406 @@ static void read_dir_cb(u16 sender_id, u8 len, u8 msg[], void *context);
...
+                    fd_cache[idx].path,
+                    fd_cache[idx].mode,
+                    fd_cache[idx].offset);
+      fclose(fd_cache[idx].fp);
+      fd_cache[idx] = empty_cache_entry;
+    }
+  }
+}
+
+static fd_cache_result_t open_from_cache(const char *path, const char *mode, int oflag, mode_t perm)
```
Comment (pmiettinen) [inline package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c]: Seeing that there is `no_cache` variable leads me to thinking is this `open_from_cache` name a bit misleading in some cases?

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c (hunk 11 above)
Comment (silverjam) [reply package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c]: 👍

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.h (hunk 10 above)
Comment (silverjam) [reply package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.h]: 👍

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c (hunk 12 above)
Comment (silverjam) [reply package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c]: Changed naming to make the code less misleading

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v2.1.0-develop-2019010420-23-ge89e8e7e

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=852238)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail"
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019010420-23-ge89e8e7e&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019010420-23-ge89e8e7e&build_type=buildroot_pr&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v2.1.0-develop-2019010420-23-ge89e8e7e
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2019010420-23-ge89e8e7e)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2019010420-23-ge89e8e7e/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2019010420-23-ge89e8e7e

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v2.1.0-develop-2019010420-25-ge2588bb9

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=852417)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail"
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019010420-25-ge2588bb9&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019010420-25-ge2588bb9&build_type=buildroot_pr&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v2.1.0-develop-2019010420-25-ge2588bb9
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2019010420-25-ge2588bb9)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2019010420-25-ge2588bb9/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2019010420-25-ge2588bb9

**Comment by silverjam** [comment]
Clang format passed, merging...

**Pull request merged**